### ISSUE TITLE
Cannot configure multiple bulk redirect list using Terraform

### ISSUE THREAD

**Issue by tchellomello** [cloudflare/terraform-provider-cloudflare#3036]
### Confirmation

- [X] My issue isn't already found on the issue tracker.
- [X] I have replicated my issue using the latest version of the provider and it is still present.

### Terraform and Cloudflare provider version

```bash
❯ terraform -v
Terraform v1.6.6
on linux_amd64
+ provider registry.terraform.io/cloudflare/cloudflare v4.21.0
+ provider registry.terraform.io/hashicorp/aws v5.31.0
```

### Affected resource(s)

```
cloudflare_ruleset
```

### Terraform configuration files

```hcl
resource "cloudflare_list" "bulk_redirect_to_id" {
  account_id  = var.cloudflare_account_id
  name        = var.bulk_redirect_list_name
  description = var.bulk_redirect_list_description
  kind        = "redirect"
}

resource "cloudflare_list_item" "bulk_redirect_to_id_item" {
   
  # the issue happens even with 1 resource
  for_each = { for redirect in var.bulk_redirects : "${redirect.source_url}" => redirect }

  account_id = var.cloudflare_account_id
  list_id    = cloudflare_list.bulk_redirect_to_id.id

  redirect {
    source_url  = each.value.source_url
    target_url  = each.value.target_url
    status_code = each.value.status_code
  }

  depends_on = [
    cloudflare_list.bulk_redirect_to_id
  ]

}

# Create Bulk Redirect and attach bulk redirect list
resource "cloudflare_ruleset" "bulk_root_redirect_to_id" {
  account_id  = var.cloudflare_account_id
  name        = var.bulk_redirect_list_name
  description = var.bulk_redirect_list_description
  kind        = "root"
  phase       = "http_request_redirect"

  rules {
    action = "redirect"
    action_parameters {
      from_list {
        name = var.bulk_redirect_list_name // name of bulk redirect list
        key  = "http.request.full_uri"
      }
    }
    expression  = "http.request.full_uri in ${"$"}${var.bulk_redirect_list_name}" // requires the $ sign
    description = var.bulk_redirect_list_description
    enabled     = true
  }

  depends_on = [
    cloudflare_list_item.bulk_redirect_to_id_item
  ]
}


## variables.tf
variable "cloudflare_account_id" {
  default = "123456"
}

variable "bulk_description" {
  default = "my desc"
}

variable "bulk_redirects" {
  type = map(object({
    source_url  = string
    target_url  = optional(string, "https://my.redir.example")
    status_code = optional(number, 301)
  }))

  default = {
    "demo"         = { source_url = "https://demo.example.com/" },
  }
}
```

### Link to debug output

https://gist.github.com/tchellomello/9fa111f814cd5e20917a5efd27e1a2ed

### Panic output

```tf
2024-01-03T17:35:13.725-0500 [ERROR] provider.terraform-provider-cloudflare_v4.21.0: Response contains error diagnostic: @caller=github.com/hashicorp/terraform-plugin-go@v0.20.0/tfprotov6/internal/diag/diagnostics.go:62 diagnostic_severity=ERROR tf_proto_version=6.4 diagnostic_summary="error creating ruleset bulk_root_id_redir" tf_provider_addr=registry.terraform.io/cloudflare/cloudflare tf_resource_type=cloudflare_ruleset tf_rpc=ApplyResourceChange @module=sdk.proto diagnostic_detail="exceeded maximum number of root rulesets" tf_req_id=946222ab-150a-212d-addc-7d5d98a8e96d timestamp=2024-01-03T17:35:13.724-0500
2024-01-03T17:35:13.725-0500 [DEBUG] State storage *remote.State declined to persist a state snapshot
2024-01-03T17:35:13.725-0500 [ERROR] vertex "module.bulk_root_redirect_to_id.cloudflare_ruleset.bulk_root_redirect_to_id" error: error creating ruleset bulk_root_id_redir
2024-01-03T17:35:13.726-0500 [DEBUG] states/remote: state read serial is: 28; serial is: 28
2024-01-03T17:35:13.726-0500 [DEBUG] states/remote: state read lineage is: 367e5b02-60c3-f659-442d-8ea9def3d1b6; lineage is: 367e5b02-60c3-f659-442d-8ea9def3d1b6
╷
│ Error: error creating ruleset bulk_root_id_redir
│ 
│   with module.bulk_root_redirect_to_id.cloudflare_ruleset.bulk_root_redirect_to_id,
│   on ../../../modules/bulk_redirects/main.tf line 29, in resource "cloudflare_ruleset" "bulk_root_redirect_to_id":
│   29: resource "cloudflare_ruleset" "bulk_root_redirect_to_id" {
│ 
│ exceeded maximum number of root rulesets
````

### Expected output

Bulk Redirect list be created. 

### Actual output

```tf
❯ terraform apply
[...]
Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  + create

Terraform will perform the following actions:

  # module.bulk_root_redirect_to_id.cloudflare_ruleset.bulk_root_redirect_to_id will be created
  + resource "cloudflare_ruleset" "bulk_root_redirect_to_id" {
      + account_id  = "81<ACCOUNT_ID>"
      + description = "Former Root Redirect"
      + id          = (known after apply)
      + kind        = "root"
      + name        = "bulk_root_id_redir"
      + phase       = "http_request_redirect"

      + rules {
          + action       = "redirect"
          + description  = "Former Root Redirect"
          + enabled      = true
          + expression   = "http.request.full_uri in $bulk_root_id_redir"
          + id           = (known after apply)
          + last_updated = (known after apply)
          + ref          = (known after apply)
          + version      = (known after apply)

          + action_parameters {
              + version = (known after apply)

              + from_list {
                  + key  = "http.request.full_uri"
                  + name = "bulk_root_id_redir"
                }
            }
        }
    }

Plan: 1 to add, 0 to change, 0 to destroy.

Do you want to perform these actions?
  Terraform will perform the actions described above.
  Only 'yes' will be accepted to approve.

  Enter a value: yes

module.bulk_root_redirect_to_id.cloudflare_ruleset.bulk_root_redirect_to_id: Creating...
╷
│ Error: error creating ruleset bulk_root_id_redir
│ 
│   with module.bulk_root_redirect_to_id.cloudflare_ruleset.bulk_root_redirect_to_id,
│   on ../../../modules/bulk_redirects/main.tf line 29, in resource "cloudflare_ruleset" "bulk_root_redirect_to_id":
│   29: resource "cloudflare_ruleset" "bulk_root_redirect_to_id" {
│ 
│ exceeded maximum number of root rulesets
```

### Steps to reproduce

1. Invoke the module
2. Issue is reproduced

### Additional factoids

Initially we were hitting the issue reported at https://github.com/cloudflare/terraform-provider-cloudflare/issues/2970 which was fixed. Now we are able to create the bulk redirect list and append the URLs on its list.

Found some other links related to a similar problem:

* https://community.cloudflare.com/t/cannot-configure-multiple-bulk-redirect-lists-using-terraform/500043/3
* https://community.cloudflare.com/t/create-bulk-redirect-rule-via-api-400-error-exceeded-maximum-number-of-root-rulesets/505154/4

### References

* https://github.com/cloudflare/terraform-provider-cloudflare/issues/2004
* https://github.com/cloudflare/terraform-provider-cloudflare/issues/2970

**Comment by jacobbednarz** [comment]
the issue you've raised here is not a provider issue but the way the rulesets engine is designed and has entitlement restrictions. you should reach out to your account team to have them address this.

**Comment by tchellomello** [comment]
> the issue you've raised here is not a provider issue but the way the rulesets engine is designed and has entitlement restrictions. you should reach out to your account team to have them address this.

@jacobbednarz Thanks for your comment. IMHO if that was something related on my account or due to entitlement restrictions, I would not be able to create/activate the ruleset via webUI, which I can. I understand that the error is coming from the API, but the fact that I can activate it via webUI seems to point to the code.

I'll be happy to raise a ticket with the support team, but the fact that I can create and enable the ruleset via webUI does change anything on your statement?

Thanks!

**Comment by jacobbednarz** [comment]
depending on what you are doing the UI will consolidate it into a single root ruleset for the phase. you can do the same thing in terraform but you'll need to fix your configuration to match that behaviour instead of what you are doing now (trying to create multiples through your module). it still isn't a bug and a restriction of ERE. full docs are available at https://developers.cloudflare.com/ruleset-engine/

**Comment by tchellomello** [comment]
@jacobbednarz just to provide closure, we were able to get this addressed. The problem was not on our Terraform code/logic but on a bad API data which we believe was a bad leftover during the attempts reported at https://github.com/cloudflare/terraform-provider-cloudflare/issues/2970. 

As mentioned on the https://github.com/cloudflare/terraform-provider-cloudflare/issues/2004, the data was not being displayed an the UI which was causing a trigger a generic API exception. 

Basically we manually forced a DELETE and then the same Terraform code worked as expected. 

```tf
❯ curl --request DELETE https://api.cloudflare.com/client/v4/accounts/XXXXXX/rulesets/5xxxxxx \
[....]
```

Thanks for your help!
